### PR TITLE
Add support for SPV_INTEL_fp_fast_math_mode

### PR DESCRIPTION
**Pull request by MochalovaAn**
Spec: https://github.com/KhronosGroup/SPIRV-Registry/pull/85

Signed-off-by: amochalo <anastasiya.mochalova@intel.com>

### CODE REVIEW

File: lib/SPIRV/SPIRVWriter.cpp [hunk 1]
```diff
@@ -1646,6 +1646,14 @@ bool LLVMToSPIRV::transDecoration(Value *V, SPIRVValue *BV) {
           M |= FPFastMathModeNSZMask;
         if (FMF.allowReciprocal())
           M |= FPFastMathModeAllowRecipMask;
+        if (FMF.allowContract()) {
+          M |= FPFastMathModeAllowContractMask;
+          BM->addCapability(CapabilityFPFastMathModeINTEL);
```
Comment (AlexeySachkov) [inline lib/SPIRV/SPIRVWriter.cpp:1651]: You should only do that if extension is enabled. Example: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/625/files#diff-ba003da77a80df1db76fb378a873a80e758591d72bf82a46fc90069b2a947990R530

File: lib/SPIRV/libSPIRV/spirv.hpp [hunk 2]
```diff
@@ -376,6 +376,8 @@ enum FPFastMathModeMask {
     FPFastMathModeNotInfMask = 0x00000002,
     FPFastMathModeNSZMask = 0x00000004,
     FPFastMathModeAllowRecipMask = 0x00000008,
+    FPFastMathModeAllowContractMask = 0x10000,
```
Comment (AlexeySachkov) [inline lib/SPIRV/libSPIRV/spirv.hpp:379]: I think it is better to add `INTEL` suffix to those to highlight that these do not belong to the code SPIR-V spec

File: test/transcoding/fp_contract_reassoc_fast_mode.ll [hunk 3]
```diff
@@ -0,0 +1,50 @@
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv -spirv-text %t.bc -o - | FileCheck %s --check-prefix=CHECK-SPIRV
```
Comment (AlexeySachkov) [inline test/transcoding/fp_contract_reassoc_fast_mode.ll:2]: Test should be adjusted to incorporate the fact that the new extension might be disabled and must be enabled in order to take effect

**Review by AlexeySachkov — COMMENTED** [review]
Please add a link to the spec and adjust PR title/description: mentioned pragma are not relevant to the SPIR-V translator at all

File: lib/SPIRV/libSPIRV/spirv.hpp (hunk 2 above)
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/spirv.hpp:379]: I also suggest to make the assigned value to be 0x00010000, so the changes would be:
```
    FPFastMathModeAllowContractFastINTELMask = 0x00010000,
    FPFastMathModeAllowReassocINTELMask = 0x00020000,
```

File: test/transcoding/fp_contract_reassoc_fast_mode.ll [hunk 4]
```diff
@@ -1,14 +1,21 @@
...
+; CHECK-SPIRV-ON: 3 Name [[su:[0-9]+]] "sub"
+; CHECK-SPIRV-ON: 4 Decorate [[mu]] FPFastMathMode 65536
+; CHECK-SPIRV-ON: 4 Decorate [[su]] FPFastMathMode 131072
 
-; CHECK-SPIRV: 2 Capability FPFastMathModeINTEL
-; CHECK-SPIRV: 3 Name [[mu:[0-9]+]] "mul"
-; CHECK-SPIRV: 3 Name [[su:[0-9]+]] "sub"
-; CHECK-SPIRV: 4 Decorate [[mu]] FPFastMathMode 65536
-; CHECK-SPIRV: 4 Decorate [[su]] FPFastMathMode 131072
+; CHECK-SPIRV-OFF-NOT: 2 Capability FPFastMathModeINTEL;
```
Comment (MrSidims) [inline test/transcoding/fp_contract_reassoc_fast_mode.ll:14]:
```suggestion
; CHECK-SPIRV-OFF-NOT: 2 Capability FPFastMathModeINTEL
```

**Review by MrSidims — APPROVED** [review]
Other than a little typo LGTM

File: test/transcoding/fp_contract_reassoc_fast_mode.ll [hunk 5]
```diff
@@ -0,0 +1,57 @@
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv -spirv-text %t.bc -o - | FileCheck %s --check-prefix=CHECK-SPIRV-OFF
+; RUN: llvm-spirv --spirv-ext=+SPV_INTEL_fp_fast_math_mode -spirv-text %t.bc -o - | FileCheck %s --check-prefix=CHECK-SPIRV-ON
+; RUN: llvm-spirv  --spirv-ext=+SPV_INTEL_fp_fast_math_mode %t.bc -o %t.spv
+; RUN: spirv-val %t.spv
```
Comment (MrSidims) [inline test/transcoding/fp_contract_reassoc_fast_mode.ll:5]: That is why the test is failing: spirv-val doesn't know about the added capability here (and PR https://github.com/KhronosGroup/SPIRV-Headers/pull/176, where the extension is added to SPIR-V Headers isn't yet merged). I suggest to remove it for now.

Comment (MrSidims) [inline test/transcoding/fp_contract_reassoc_fast_mode.ll:5]:
```suggestion
```

**Review by MrSidims — APPROVED** [review]

**Review by AlexeySotkin — APPROVED** [review]

File: test/transcoding/fp_contract_reassoc_fast_mode.ll [hunk 6]
```diff
@@ -0,0 +1,56 @@
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv -spirv-text %t.bc -o - | FileCheck %s --check-prefix=CHECK-SPIRV-OFF
+; RUN: llvm-spirv --spirv-ext=+SPV_INTEL_fp_fast_math_mode -spirv-text %t.bc -o - | FileCheck %s --check-prefix=CHECK-SPIRV-ON
+; RUN: llvm-spirv  --spirv-ext=+SPV_INTEL_fp_fast_math_mode %t.bc -o %t.spv
```
Comment (mlychkov) [inline test/transcoding/fp_contract_reassoc_fast_mode.ll:4]: Please remove extra space.

**Pull request merged**